### DISCUSSION TITLE
com.fasterxml.jackson.databind from 2.10.1 to 2.13.3

### DISCUSSION THREAD

**Discussion by Odoiporos** [FasterXML/jackson-databind#4730] [Q&A]
Hi,

Until recently we had the com.fasterxml.jackson.databind from 2.10.1 version and for an objectMapper.convertValue to a dto that has an enum property we had a custom method annotated with @jsonCreator in order to work properly

when we change dto 2.13.3 we had to use the (mode = JsonCreator.Mode.DELEGATING)
Can you please explain what has changed between theses versions that makes the above mode necessary?

**Comment by pjfanning** [comment]
Please provide a full example. These are not even recent releases. Without providing some real detail, it will be close to impossible to answer this.

**Comment by Odoiporos** [comment]
the error is 
```
java.lang.IllegalArgumentException: Input mismatch reading Enum `***.AltamiraApplicationStatus`: properties-based `@JsonCreator` ([method ***.AltamiraApplicationStatus#fromAltamiraValue(java.lang.String)]) expects JSON Object (JsonToken.START_OBJECT), got JsonToken.VALUE_STRING
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: ***.AccountBeneficiaryApplicationAltamiraResponse["applicationStatus"])
```

the enum itself 
```
@Getter
@RequiredArgsConstructor
public enum AltamiraApplicationStatus {
    INITIALIZED("0"), 
    SUBMITTED("1"), 
    PENDING_BENEFICIARIES("2"), 
    PENDING_CANDIDATE("3"), 
    ACCEPTED("4"), 
    REJECTED("5"),
    EXPIRED("6"),
    COMPLETED("7"),
    FAILED("8"),
    FAILED_ACC_CLOSED("9");

    private final String code;

    @JsonCreator(mode = JsonCreator.Mode.DELEGATING)
    public static AltamiraApplicationStatus fromAltamiraValue(String code) {
        return Stream.of(AltamiraApplicationStatus.values())
                .filter(d -> d.getCode().equalsIgnoreCase(code) || d.toString().equals(code))
                .findFirst()
                .orElse(null);
    }
}
```

**Reply by pjfanning** [reply]
And what happens if you use a new release (2.18.0, 2.17.2) ?

**Comment by JooHyukKim** [comment]
Can you check the release notes, in [FasterXML/Jackson wiki](https://github.com/FasterXML/jackson/wiki), @Odoiporos ? There has been a lot of change between before and after versions. 

And yes like @pjfanning said, try out later versions. 
2.13 is no longer maintained version, so we won't get it fixed in 2.13 even if we do find the root cause.

**Comment by Odoiporos** [comment]
ok thank you both for your answers.
i will check with newer versions if the mode is mandatory. i mostly wanted to know if that change was on purpose.

**Reply by cowtowncoder** [reply]
Your faith in our knowledge of all changes across versions is humbling... :)

Since this is within not super recent versions, the answers is that:

1. Almost certainly not intentional wrt `convertValue()`, as that is not something with which `@JsonCreator` and other annotations are tested
2. Likely intentional change in itself, to handle some issue with "regular" use of `@JsonCreator` wrt `ObjectMapper.readValue()` and so on.

Use of `mode = JsonCreator.Mode.DELEGATING` is pretty typical for 1-arg case because of its inherent ambiguity (is it single property we are binding out of JSON Object -- OR full JSON value?).

But the real thing wrt compatibility is how do 2.17.2 and 2.18.0 function -- that is likely the current best approximation of expected behavior.
